docs/combininglock.md

```diff
@@ -0,0 +1,342 @@
+The `ACQUIRE` part is almost the same as the MCS queue lock.
+However, it additionally checks if another thread completed the lambda for this thread.
+If this happens, then the thread can return immediately.
+
+The `PERFORM` part is where the executing thread can run the lambdas for itself and many other threads.
+It walks the queue executing the lambdas until it reaches a queue node with `next` pointer containing `nullptr`.
+
+The `RELEASE` part attempts to remove the last node from the `PERFORM` phase from the end of the queue.
+If this succeeds, then it can simply notify that queue node that its work is `DONE`,
+and return.
+Otherwise, as with the MCS lock, it must wait for a new thread that is joining the queue to set the `next` pointer.
+Then it can notify (`HEAD`) the newly joined thread that it can execute the remaining work.
+Finally, it must notify the last element it ran the lambda for that its work is `DONE`.
+The `DONE` notification must be done after the read of `next`, as signalling `DONE` can cause the memory for the node to be freed from the stack.
+
+The [actual implementation](../src/snmalloc/ds/combininglock.h) in the `snmalloc` codebase is a little more complex as it uses a fast path lock for the uncontended case.
+This uses fewer atomic operations in the common case where the lock is not contended,
+and is a standard technique for queue locks.
+
+
+## Performance Results
+
+We have a particularly tough benchmark for testing [lock contention at startup](../src/test/perf/startup/startup.cc).
+We used a machine with 72 hardware threads.
+The benchmark causes all the threads to synchronise on starting their first allocation.
+This means all 72 threads are contending on the lock at the same time to get their allocator initialised.
+
+We ran this bench mark with a standard spin-lock (0.7.0-spin) and with the new combining lock (0.7.0):
+
+![Combining lock performance](combininglockperf.svg)
+
+As we can see from the benchmark, the combining lock is significantly faster than the spin lock in this highly contented scenario taking only 60% of the time to complete the initialisation.
+
+This benchmark is not at all representative of most scenarios, and is stressing a worst case scenario of the system.
+
+## Future Work
+
+There are two things that would make the Combining lock even better:
+
+* Back off strategies
+* Integration with Futex/Waitonaddress
+
+The current implementation does not have any back off strategies for its spinning loops.
+This can be bad for performance, however, with the MCS queue lock the threads spin on individual cache lines so some of the worst effects of [spinning are mitigated](https://pdos.csail.mit.edu/papers/linux:lock.pdf).
+However, for a more general purpose the combining lock should have a back off strategy to increase the time between checks the longer it waits.
+
+The second improvement would be to integrate the combining lock with the OS level support for waiting.
```

Comment:
What Rust's synchronization routines are generally using is to spin `100` (a somehow chosen number) times before going into sleep:
```c++
      int remaining_spins = SPIN_COUNT;
      // Do spin polling for certain amount of time.
      while (remaining_spins > 0) {
        if (header.status.load(cpp::MemoryOrder::RELAXED) !=
            RawLambdaLockHeader::WAITING)
          break;
        sleep_briefly();
        remaining_spins--;
      }
      // If we used up all spins, we may need to go to sleep.
      if (remaining_spins == 0) {
        FutexWordType expected = RawLambdaLockHeader::WAITING;
        if (header.status.compare_exchange_strong(
                expected, RawLambdaLockHeader::SLEEPING,
                cpp::MemoryOrder::ACQ_REL, cpp::MemoryOrder::RELAXED))
          header.status.wait(RawLambdaLockHeader::SLEEPING);
      }
```
However, that was for polling shared words. Not sure if exponential backoff or other schemes would become more applicable in MCS lock.